### ISSUE TITLE
🐛 [Extension] Biome LSP Server crashes in Cursor editor

### ISSUE THREAD

**Issue by TheRealMentorV2** [biomejs/biome#7586]
### Environment information

```block
CLI:
  Version:                      2.0.6
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v18.18.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         npm/9.8.1

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true

Workspace:
  Open Documents:               0
```

### What happened?

I am trying to integrate biome to a monorepo. All the scripts and packages are installed and it works properly if ran on CLI. But when I use biome extension in cursor, it shows me following error:

```
2025-09-25 01:40:41.894 [info] Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue:

Source Location: crates/biome_service/src/settings.rs:78:14
Thread Name: tokio-runtime-worker
Message: You must have at least one workspace.


2025-09-25 01:40:41.898 [info] [Error - 01:40:41] Server process exited with code 0. 
```

### Expected result

It should run without crashing.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
Please use the latest. Let us know if that works and fixes the issue. And please provide a reproduction

**Comment by ghost** [comment]
Hi @ematipico can use this repo to reproduce the issue -- https://github.com/TheRealMentorV2/biome-issue-7586

**Useful Information**

- Biome Extension Version: `3.2.0` (latest)
- Biome CLI version: `2.2.4` (latest)
- `assist` doesn't seem to be supported in `2.2.4` and gives error when added to `biome.json` but it is mentioned in the docs that it is supported in v2.

**How I reproduced the issue?**

- Init the repo using CLI - `npm create @biomejs/biome-reproduction`
- Added dummy code with formatting issues (check line 25 in `src/index.ts`)
- Run `pnpm exec biome check` to verify if biome is working (no issues)
- Add `assist` in `biome.json`
- Biome LSP crashes

<img width="1256" height="306" alt="Image" src="https://github.com/user-attachments/assets/1d46c32a-2aac-4e88-aaeb-448ca4440d8d" />
<img width="557" height="202" alt="Image" src="https://github.com/user-attachments/assets/543c274f-4d2f-43da-a38b-73bfe085c95a" />

**Comment by ematipico** [comment]
Is it possible you have multiple processes `biome` running in the background?

**Comment by ghost** [comment]
Yes, it seems I have multiple `biome` processes in background. Can i double check these 3 that i see on Cursor terminal are expected?

<img width="220" height="65" alt="Image" src="https://github.com/user-attachments/assets/673a7736-4799-4919-9865-c505f1172724" />

**Comment by ematipico** [comment]
Close the editor, kill the ghost processes with a task manager, open the editor. See if that fixes it. We already have an issue open for those multiple processes

**Comment by ghost** [comment]
No, it doesn't seem to work. Crashed again. And this time, there's only 1 biome process running.

```
2025-09-25 13:15:20.669 [info] [Info  - 13:15:20] Connection to server got closed. Server will restart.
2025-09-25 13:15:20.669 [info] true
2025-09-25 13:15:20.671 [info] [Error - 13:15:20] Server process exited with code 0.
2025-09-25 13:15:20.799 [info] Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue:

Source Location: crates/biome_service/src/settings.rs:78:14
Thread Name: tokio-runtime-worker
Message: You must have at least one workspace.
```

Can you share the link to that open issue so I can also follow up? Thanks~

**Comment by ghost** [comment]
Hi @ematipico any updates regarding this issue?

**Comment by wojtekmaj** [comment]
For me, the latest version of Biome is crashing ALL THE TIME on VSCode, rendering it _completely_ unusable. But I guess since Cursor is a fork of VSCode there's no need to file another issue for that.

### TL;DR

There's a regression introduced in v2.2.3 that causes Biome LSP to crash on macOS.

### Test

I did a simple test:
* Added an extra new line that Biome formatter would surely remove
* Installed a particular Biome version (see below)
* Checked if it will remove the extra line on save

On the latest versions of Biome, it just crashes the process.
I can reproduce the issue on:
- v2.2.4
- v2.2.3

Immediately after rolling back to any of these earlier versions, saving the file removes the extra line and the process does not crash.
I CANNOT reproduce the issue on:
- v2.2.2
- v2.2.0
- v2.0.0

Other factors that do make a difference:
- Works on Windows (arm64) on all versions, actually
- Does not work on macOS

Other factors that do NOT seem to make any difference:
- Using pre-release Biome VSCode extension vs using the latest stable
- Having code assist or not in biome.json
- Whether it's a multi-root workspace or not

### Logs

Biome extension logs:
```
2025-09-30 10:34:05.451 [info]  🚀 Biome extension 2025.7.41733.
2025-09-30 10:34:05.451 [info]  ✨ Running in single-root workspace mode.
2025-09-30 10:34:05.451 [info]  🔍 Found 1 workspace folder(s).
2025-09-30 10:34:05.451 [info]  🚀 Created 1 Biome instance(s).
2025-09-30 10:34:05.495 [info]  🚀 Started 1 Biome instance(s).
2025-09-30 10:34:05.495 [info]  ✨ See the dedicated logging channels for more details.
2025-09-30 10:36:24.262 [info]   ---------------------------------------- 
```

Biome instance logs:
```
2025-09-30 10:34:05.451 [info]  🔒 Started listening for lockfile changes.
2025-09-30 10:34:05.464 [info]  ⚙️ Started listening for extension settings changes.
2025-09-30 10:34:05.495 [info]  ✅ Biome is ready.
```

Biome LSP logs:
```
2025-09-30 10:34:05.533 [info] [Info  - 10:34:05] Server initialized with PID: 63044
2025-09-30 10:34:18.945 [info] [Info  - 10:34:18] Connection to server got closed. Server will restart.
2025-09-30 10:34:18.945 [info] true
2025-09-30 10:34:18.948 [info] [Error - 10:34:18] Server process exited with code 0.
2025-09-30 10:34:19.205 [info] [Error - 10:34:19] Server process exited with code 0.
2025-09-30 10:34:19.206 [info] [Info  - 10:34:19] Connection to server got closed. Server will restart.
2025-09-30 10:34:19.206 [info] true
2025-09-30 10:34:22.364 [info] [Error - 10:34:22] Server process exited with code 0.
2025-09-30 10:34:22.364 [info] [Info  - 10:34:22] Connection to server got closed. Server will restart.
2025-09-30 10:34:22.364 [info] true
2025-09-30 10:34:22.390 [info] [Error - 10:34:22] Server process exited with code 0.
2025-09-30 10:34:22.390 [info] [Info  - 10:34:22] Connection to server got closed. Server will restart.
2025-09-30 10:34:22.390 [info] true
2025-09-30 10:34:22.494 [info] [Error - 10:34:22] Server process exited with code 0.
2025-09-30 10:34:22.494 [info] [Error - 10:34:22] The biome server crashed 5 times in the last 3 minutes. The server will not be restarted. See the output for more information.
```

### Suspect 
> [!WARNING]
> Insane AI slop ahead.

AI suspects this PR: https://github.com/biomejs/biome/pull/7377/files as the most plausible cause for the crash described above.

> The crash is likely caused by the changes in did_open:
> - If project_for_path(&path) is None, the handler now just logs and returns early.
> - That means the file is never registered (Document missing, no project_key).
> - Later LSP handlers (didChange, formatting, semantic tokens, diagnostics) assume the document/project exists → panic or server crash.
> - Previously, did_open would try load_biome_configuration_file(...) as a fallback, which avoided this.
> 
> So the regression is: files opened before a project is fully loaded leave the server in an inconsistent state.
> 
> Fix ideas:
> - Reintroduce a fallback project load in did_open when project_for_path is None, or
> - At least register a stub Document so later handlers no-op safely instead of crashing.

**Comment by ghost** [comment]
Thanks @wojtekmaj let me give it a try with the versions you mentioned above. I will get back to you guys soon.

**Comment by ZlatanCN** [comment]
> For me, the latest version of Biome is crashing ALL THE TIME on VSCode, rendering it _completely_ unusable. But I guess since Cursor is a fork of VSCode there's no need to file another issue for that.
> 
> ### TL;DR
> There's a regression introduced in v2.2.3 that causes Biome LSP to crash on macOS.
> 
> ### Test
> I did a simple test:
> 
> * Added an extra new line that Biome formatter would surely remove
> * Installed a particular Biome version (see below)
> * Checked if it will remove the extra line on save
> 
> On the latest versions of Biome, it just crashes the process. I can reproduce the issue on:
> 
> * v2.2.4
> * v2.2.3
> 
> Immediately after rolling back to any of these earlier versions, saving the file removes the extra line and the process does not crash. I CANNOT reproduce the issue on:
> 
> * v2.2.2
> * v2.2.0
> * v2.0.0
> 
> Other factors that do make a difference:
> 
> * Works on Windows (arm64) on all versions, actually
> * Does not work on macOS
> 
> Other factors that do NOT seem to make any difference:
> 
> * Using pre-release Biome VSCode extension vs using the latest stable
> * Having code assist or not in biome.json
> * Whether it's a multi-root workspace or not
> 
> ### Logs
> Biome extension logs:
> 
> ```
> 2025-09-30 10:34:05.451 [info]  🚀 Biome extension 2025.7.41733.
> 2025-09-30 10:34:05.451 [info]  ✨ Running in single-root workspace mode.
> 2025-09-30 10:34:05.451 [info]  🔍 Found 1 workspace folder(s).
> 2025-09-30 10:34:05.451 [info]  🚀 Created 1 Biome instance(s).
> 2025-09-30 10:34:05.495 [info]  🚀 Started 1 Biome instance(s).
> 2025-09-30 10:34:05.495 [info]  ✨ See the dedicated logging channels for more details.
> 2025-09-30 10:36:24.262 [info]   ---------------------------------------- 
> ```
> 
> Biome instance logs:
> 
> ```
> 2025-09-30 10:34:05.451 [info]  🔒 Started listening for lockfile changes.
> 2025-09-30 10:34:05.464 [info]  ⚙️ Started listening for extension settings changes.
> 2025-09-30 10:34:05.495 [info]  ✅ Biome is ready.
> ```
> 
> Biome LSP logs:
> 
> ```
> 2025-09-30 10:34:05.533 [info] [Info  - 10:34:05] Server initialized with PID: 63044
> 2025-09-30 10:34:18.945 [info] [Info  - 10:34:18] Connection to server got closed. Server will restart.
> 2025-09-30 10:34:18.945 [info] true
> 2025-09-30 10:34:18.948 [info] [Error - 10:34:18] Server process exited with code 0.
> 2025-09-30 10:34:19.205 [info] [Error - 10:34:19] Server process exited with code 0.
> 2025-09-30 10:34:19.206 [info] [Info  - 10:34:19] Connection to server got closed. Server will restart.
> 2025-09-30 10:34:19.206 [info] true
> 2025-09-30 10:34:22.364 [info] [Error - 10:34:22] Server process exited with code 0.
> 2025-09-30 10:34:22.364 [info] [Info  - 10:34:22] Connection to server got closed. Server will restart.
> 2025-09-30 10:34:22.364 [info] true
> 2025-09-30 10:34:22.390 [info] [Error - 10:34:22] Server process exited with code 0.
> 2025-09-30 10:34:22.390 [info] [Info  - 10:34:22] Connection to server got closed. Server will restart.
> 2025-09-30 10:34:22.390 [info] true
> 2025-09-30 10:34:22.494 [info] [Error - 10:34:22] Server process exited with code 0.
> 2025-09-30 10:34:22.494 [info] [Error - 10:34:22] The biome server crashed 5 times in the last 3 minutes. The server will not be restarted. See the output for more information.
> ```
> 
> ### Suspect
> Warning
> 
> Insane AI slop ahead.
> 
> AI suspects this PR: [#7377 (files)](https://github.com/biomejs/biome/pull/7377/files) as the most plausible cause for the crash described above.
> 
> > The crash is likely caused by the changes in did_open:
> > 
> > * If project_for_path(&path) is None, the handler now just logs and returns early.
> > * That means the file is never registered (Document missing, no project_key).
> > * Later LSP handlers (didChange, formatting, semantic tokens, diagnostics) assume the document/project exists → panic or server crash.
> > * Previously, did_open would try load_biome_configuration_file(...) as a fallback, which avoided this.
> > 
> > So the regression is: files opened before a project is fully loaded leave the server in an inconsistent state.
> > Fix ideas:
> > 
> > * Reintroduce a fallback project load in did_open when project_for_path is None, or
> > * At least register a stub Document so later handlers no-op safely instead of crashing.

thank you @wojtekmaj , v2.2.2 works for me

**Comment by bmatzner** [comment]
Can confirm this with biome >2.2.2 (up to currently 2.2.6) in a large workspace with different repos with their own biome.json configs, in Windsurf Mac.

**Comment by ghost** [comment]
> Thanks [@wojtekmaj](https://github.com/wojtekmaj) let me give it a try with the versions you mentioned above. I will get back to you guys soon.

I tried with biome v2.2.2 and biome extension v2.2.2, no issues now. 

If the issue is already being followed in another thread, then I think we can close this issue for now.

**Comment by itajenglish** [comment]
+1 Hope this gets fixed soon

**Comment by CarlosZiegler** [comment]
Hey updated to latest version and got same error

**Comment by tomerzcod7** [comment]
Any update on this? happens to all of us

**Comment by siketyan** [comment]
I couldn't reproduce the issue with @TheRealMentorV2 's reproduction. I tried both VS Code and Cursor, but it doesn't reproduce either. Anyone please provide a valid, minimal reproduction or we can't do much. Make sure the reproduction contains all your VS Code / Cursor settings.
